File: Chapter 03/JS/3.3.4 ISP/Before/midfielder.js
// JavaScript doesn't have native Interfaces, but we simulate the architectural behavior by defining the classes exactly 
// how they look conceptually!
class Midfielder {
    practiceShooting() {
        console.log("  [Midfielder] Practicing shooting drills.");
    }

    practiceTackling() {
        console.log("  [Midfielder] Practicing slide tackles.");
    }

    // Forced to include these by the "Fat Interface" design
    practiceDivingSaves() {
        throw new Error("Midfielders don't play in the net!");
    }

    practiceHandDistribution() {
        throw new Error("Midfielders can't use their hands!");
    }
}

module.exports = Midfielder;

File: Chapter 03/C#/3.3.4 ISP/Before/Midfielder.cs
using System;

namespace Chapter03.ISP.Before;

public class Midfielder : ITrainingSession
{
    public void PracticeShooting() 
    { 
        Console.WriteLine("  [Midfielder] Practicing shooting drills."); 
    }
    
    public void PracticeTackling() 
    { 
        Console.WriteLine("  [Midfielder] Practicing slide tackles."); 
    }
    
    public void PracticeDivingSaves() 
    {
        throw new NotImplementedException("Midfielders don't play in the net!");
    }
    
    public void PracticeHandDistribution() 
    {
        throw new NotImplementedException("Midfielders can't use their hands!");
    }
}